### DISCUSSION TITLE
Github Actions Runners cache in Container

### DISCUSSION THREAD

**Discussion by MarleDK** [ubicloud/ubicloud#1930] [Q&A]
I am running into an issue using the ubicloud/cache, when I have a job in GHA, which uses a container (on the job, not on each step), the cache step generates the following error, and it counts as a cache miss:

> Warning: An internal error has occurred in cache backend. Please check https://www.githubstatus.com/ for any ongoing issue in actions.

From looking at the ubicloud/cache code, and some debugging, it seems to be because ubicloud adds some environment variables: `UBICLOUD_CACHE_URL` and `UBICLOUD_RUNTIME_TOKEN`, which are not available in the docker container.
The reason I think this, is that I tried to set `UBICLOUD_CACHE_URL`, and then got another error:

> Warning: Failed to restore: getCacheEntry failed: Invalid URL

Does anyone have any idea how I can get the environment variables into the container running?
I have tried adding:
```
job_name:
  container:
    options: "--env UBICLOUD_CACHE_URL=$UBICLOUD_CACHE_URL"
```

But it doesn't seem to correctly get the environment variable from the runner.

And I don't think I can use the `env:` yaml block, as the `UBICLOUD_CACHE_URL` environment variable is not accessible in the workflow in any way.

I'm thinking it could be possible by adding an output from a step potentially, but then I would have to overwrite a single step in a job, to not run in a container, and I don't really see any way that is possible either :exploding_head: 

Any suggestions are very welcome, and I hope we can find a good solution which can potentially be added as a workaround for the docs as well.

**Comment by MarleDK** [comment]
Looking a bit more into this, it seems like it would just require some changes [here](https://github.com/ubicloud/runner/blob/a77fe8a53fe9fc2a77ae4f511d46d04197bdde66/src/Runner.Worker/Handlers/ContainerActionHandler.cs#L219) in the `runner` code, but since Ubicloud hasn't made any changes to the runner, I don't know if you are using your own, and it seems like there are other environment variables that could also be added.

**Comment by ozgune** [comment]
^ @velioglu or @enescakir to reply

**Comment by velioglu** [comment]
Hey @MarleDK,

As you mentioned, ubicloud/cache introduces two environment variables that are necessary for its use. To access them within a container, you can define an initial job running on the VM and set those environment variables as outputs of that job. Then, you can pass the outputs from the first job to the second one, which is essentially what you're doing with your test.

Please note that you can obtain the UBICLOUD_CACHE_URL directly as an output from the first job. However, the UBICLOUD_RUNTIME_TOKEN cannot be passed as an output because it appears to be a secret, and GitHub does not allow passing such values between jobs. The suggested way on GitHub for handling such cases is to store the secret in a secure store and access it in the second job. You can refer to this example for guidance: [GitHub Workflow Commands - Masking and Passing a Secret Between Jobs or Workflows.](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#example-masking-and-passing-a-secret-between-jobs-or-workflows) Would it be acceptable for you to define an additional job?

We are also working on supporting cache via our infrastructure transparently. This means you won't need to update your workflow at all. You can continue using actions/cache, but under the hood, it will utilize our infrastructure to store the cached files. I'll share an update here once it is announced.

**Reply by skovhus** [reply]
Using the following steps yields different results if the job is using a container or not. 

When not using a container, the cache is used, but for a container-based job (e.g. using `container: node:22.14`), we get the error `Warning: The runner was not able to contact the cache service. Caching will be skipped`

```yaml
- uses: pnpm/action-setup@v4 
  with:
    version: 10.4.1
- uses: ubicloud/setup-node@v4
  with:
    node-version: "22.14.0"
    cache: "pnpm"
```

**Reply by velioglu** [reply]
Hi @skovhus, you can use transparent cache https://www.ubicloud.com/docs/github-actions-integration/ubicloud-cache#ubicloud-transparent-cache which is expected to work same with and without using containers. Please let us know if it also causes different results on your tests.

**Reply by skovhus** [reply]
Transparent caching seems to work nicely. There seems to be a lot of variance in how fast things can be restored from the cache, but I don't see `Warning: The runner was not able to contact the cache service.` anymore. Let us know if there is anything we can do to speed up the cache in general.
